jd4/case.py

```diff
@@ -256,20 +256,53 @@ def read_legacy_cases(config, open):
                           memory_bytes,
                           int(score_str))
 
-def read_yaml_cases(config, open):
-    for case in yaml.safe_load(config)['cases']:
-        if 'judge' not in case:
-            yield DefaultCase(partial(open, case['input']),
-                              partial(open, case['output']),
-                              parse_time_ns(case['time']),
-                              parse_memory_bytes(case['memory']),
-                              int(case['score']))
-        else:
-            yield CustomJudgeCase(partial(open, case['input']),
+def read_auto_cases(open, zip_file, time_limit='1s', memory_limit='128m'):
+    cnt = 0
+    files = zip_file.namelist()
+    prefix = ''
+    cases = []
+    re_input = re.compile(r'([a-zA-Z]*)([0-9]*)\.in')
+    for i in files:
+        match = re_input.match(i)
+        if match:
+            prefix = match.group(1)
+            cases.append(match.group(2))
+            cnt += 1
+    if cnt == 0:
+        raise FormatError('No testdata found.')
+    cases.sort()
+    for i in cases:
+        yield DefaultCase(partial(open, prefix + str(i) + '.in'),
+                          partial(open, prefix + str(i) + '.out'),
+                          parse_time_ns(time_limit),
+                          parse_memory_bytes(memory_limit),
+                          int(100//cnt))
+
+def read_yaml_cases(config, open, zip_file):
+    try:
+        cfg = yaml.safe_load(config)
+    except:
+        raise FormatError('Invalid config file')
+    if 'cases' not in cfg:
```

Comment:
How about just throw an exception and catch it in `read_cases`? Anyhow the exception caught will be passed and fallback to `read_auto_cases`.

Reply:
I think we should never fallback to auto cases once config file exists.

Reply:
A existing config file means user doesn't want to use auto cases.
This may make user harder to find the problem.

Reply:
> A existing config file means user doesn't want to use auto cases.
> This may make user harder to find the problem.

So it should throw exception instead of returning `read_auto_cases`.

Reply:
That tells user "Hey, I see you want to use a config file but you wrote a wrong one!"

Reply:
FormatError won't be caught by read_cases.
It will give SYSTEM_ERROR to the front end,telling user "FormatError('Invalid config file')".

Reply:
Actually I want to write something that can edit config.yaml at the front end.
Sometimes you have to upload the whole test cases again just because you have written a wrong config file.

Reply:
So where is it?

I mean, if there is no `cfg['case']`, you made it returning an auto generated one.

Reply:
I misunderstood.

Reply:
This isn't wrong.
Users can write like this to change the default settings:
```
memory: 64m
```
(When the want to limit memory usage to 64MiB)
or
```
time: 5s
```
(when they want to give some extra time)